File: yolov8/ZeticMLangeYoloV8-Android/app/src/main/java/com/example/zeticmlangeyolov8androidjava/MainActivity.java
package com.example.zeticmlangeyolov8androidjava;

import android.Manifest;
import android.content.Context;
import android.content.pm.PackageManager;
import android.content.res.AssetManager;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.ImageFormat;
import android.graphics.Matrix;
import android.graphics.SurfaceTexture;
import android.hardware.camera2.CameraAccessException;
import android.hardware.camera2.CameraCaptureSession;
import android.hardware.camera2.CameraCharacteristics;
import android.hardware.camera2.CameraDevice;
import android.hardware.camera2.CameraManager;
import android.hardware.camera2.CaptureRequest;
import android.hardware.camera2.params.StreamConfigurationMap;
import android.media.Image;
import android.media.ImageReader;
import android.os.Bundle;
import android.os.Handler;
import android.os.HandlerThread;
import android.util.DisplayMetrics;
import android.view.Surface;
import android.view.TextureView;
import android.widget.ImageView;

import androidx.activity.EdgeToEdge;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;

import com.zetic.ZeticMLange.ZeticMLangeDataUtils;
import com.zetic.ZeticMLange.ZeticMLangeException;
import com.zetic.ZeticMLange.ZeticMLangeModel;
import com.zetic.ZeticMLangeFeature.ZeticMLangeFeatureYolov8;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.ByteBuffer;
import java.util.Arrays;

public class MainActivity extends AppCompatActivity implements TextureView.SurfaceTextureListener {

    private final static String TAG = "ZETIC.MLange Android(Java) sample app - YoloV8";
    private ImageView imageView;
    private TextureView textureView;
    private CameraDevice cameraDevice;
    private CameraCaptureSession cameraCaptureSession;
    private ImageReader imageReader;
    private Handler backgroundHandler;
    private HandlerThread backgroundThread;

    // ZETIC.MLange related variables
    ZeticMLangeModel zeticMLangeYoloVModel;
    ZeticMLangeFeatureYolov8 zeticMLangeFeatureYolov8;

    // YAML file to initialize yolo variables
    final String cocoYamlSamplePath = "coco.yaml";

    // ZETIC.MLange function (1) - Initialization
    private void initZeticMLangeYolov8() {
        try {
            if (zeticMLangeYoloVModel == null) {
                zeticMLangeYoloVModel = new ZeticMLangeModel(this, "yolo-v8n-test");
            }

            File cocoYamlFile = new File(getFilesDir(), cocoYamlSamplePath);
            String cocoYamlPath = cocoYamlFile.getAbsolutePath();
            if (zeticMLangeFeatureYolov8 == null) {
                zeticMLangeFeatureYolov8 = new ZeticMLangeFeatureYolov8(cocoYamlPath);
            }

        } catch (ZeticMLangeException e) {
            throw new RuntimeException(e);
        }
    }

    // ZETIC.MLange function (2) - Run
    private Bitmap runZeticMLangeYolov8(Bitmap bitmap) {
        float[] floatInput = zeticMLangeFeatureYolov8.preprocess(bitmap);
        float[][] floatInputs = {floatInput};
        ByteBuffer[] inputs = ZeticMLangeDataUtils.convertFloatArrayToByteBufferArray(floatInputs);
        try {
            zeticMLangeYoloVModel.run(inputs);
            ByteBuffer[] outputs = zeticMLangeYoloVModel.getOutputBuffers();
            float[] outputFloatArray = ZeticMLangeDataUtils.convertByteBufferToFloatArray(outputs[0]);
            return zeticMLangeFeatureYolov8.postprocess(outputFloatArray);
        } catch (ZeticMLangeException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        DisplayMetrics displayMetrics = new DisplayMetrics();
        getWindowManager().getDefaultDisplay().getMetrics(displayMetrics);
        _prepareCocoYamlFromAsset();

        // [ZETIC.MLange] - (1) Initialize model and feature extractor
        initZeticMLangeYolov8();

        // Yolo test
        imageView = findViewById(R.id.imageView);
        textureView = findViewById(R.id.textureView);
        textureView.setSurfaceTextureListener(this);
    }

    private void _prepareCocoYamlFromAsset() {
        try {
            copyFileFromAssetsToData(this, cocoYamlSamplePath);

        } catch (IOException e) {
            e.printStackTrace();
        }
    }


    @Override
    public void onSurfaceTextureAvailable(SurfaceTexture surface, int width, int height) {
        openCamera();
    }

    @Override
    public void onSurfaceTextureSizeChanged(SurfaceTexture surface, int width, int height) {
    }

    @Override
    public boolean onSurfaceTextureDestroyed(SurfaceTexture surface) {
        return true;
    }

    @Override
    public void onSurfaceTextureUpdated(SurfaceTexture surface) {
    }

    private void openCamera() {
        CameraManager manager = (CameraManager) getSystemService(Context.CAMERA_SERVICE);
        try {
            String cameraId = manager.getCameraIdList()[0];

            CameraCharacteristics characteristics = manager.getCameraCharacteristics(cameraId);
            StreamConfigurationMap map = characteristics.get(CameraCharacteristics.SCALER_STREAM_CONFIGURATION_MAP);

            if (ActivityCompat.checkSelfPermission(this, Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
                return;
            }
            manager.openCamera(cameraId, stateCallback, null);

        } catch (CameraAccessException e) {
            e.printStackTrace();
        }

    }

    private final CameraDevice.StateCallback stateCallback = new CameraDevice.StateCallback() {
        @Override
        public void onOpened(@NonNull CameraDevice camera) {
            cameraDevice = camera;
            createCameraPreviewSession();
        }

        @Override
        public void onDisconnected(@NonNull CameraDevice camera) {
            cameraDevice.close();
        }

        @Override
        public void onError(@NonNull CameraDevice camera, int error) {
            cameraDevice.close();
            cameraDevice = null;
        }
    };

    private void createCameraPreviewSession() {
        try {
            final float SCALE_VAL_TO_HIDE_TEXTURE_VIEW = 0.2f;
            final int OFFVAL_TO_HIDE_TEXTURE_VIEW = 0;

            SurfaceTexture texture = textureView.getSurfaceTexture();

            int width = textureView.getWidth();
            int height = textureView.getHeight();

            Matrix txform = new Matrix();
            textureView.getTransform(txform);
            txform.setScale((float) SCALE_VAL_TO_HIDE_TEXTURE_VIEW, (float) SCALE_VAL_TO_HIDE_TEXTURE_VIEW);

            txform.postTranslate(OFFVAL_TO_HIDE_TEXTURE_VIEW, OFFVAL_TO_HIDE_TEXTURE_VIEW);
            textureView.setTransform(txform);

            Surface surface = new Surface(texture);

            imageReader = ImageReader.newInstance(width, height, ImageFormat.JPEG, 2);
            imageReader.setOnImageAvailableListener(onImageAvailableListener, backgroundHandler);

            final CaptureRequest.Builder captureRequestBuilder = cameraDevice.createCaptureRequest(CameraDevice.TEMPLATE_PREVIEW);
            captureRequestBuilder.addTarget(surface);
            captureRequestBuilder.addTarget(imageReader.getSurface());

            cameraDevice.createCaptureSession(Arrays.asList(surface, imageReader.getSurface()),
                    new CameraCaptureSession.StateCallback() {
                        @Override
                        public void onConfigured(@NonNull CameraCaptureSession session) {
                            if (cameraDevice == null) {
                                return;
                            }
                            cameraCaptureSession = session;
                            try {
                                captureRequestBuilder.set(CaptureRequest.CONTROL_AF_MODE, CaptureRequest.CONTROL_AF_MODE_CONTINUOUS_PICTURE);
                                cameraCaptureSession.setRepeatingRequest(captureRequestBuilder.build(), null, backgroundHandler);
                            } catch (CameraAccessException e) {
                                e.printStackTrace();
                            }
                        }

                        @Override
                        public void onConfigureFailed(@NonNull CameraCaptureSession session) {
                        }
                    }, null);
        } catch (CameraAccessException e) {
            e.printStackTrace();
        }
    }

    private final ImageReader.OnImageAvailableListener onImageAvailableListener = new ImageReader.OnImageAvailableListener() {
        @Override
        public void onImageAvailable(ImageReader reader) {
            Image image = reader.acquireLatestImage();
            if (image == null) {
                return;
            }
            Bitmap bitmap = rotateBitmap(imageToBitmap(image), 90);

            image.close();

            // [ZETIC.MLange] - (1) Process the bitmap image
            Bitmap proc_image = runZeticMLangeYolov8(bitmap);

            runOnUiThread(new Thread(new Runnable() {
                @Override
                public void run() {
                    imageView.setImageBitmap(proc_image);
                }
            }));
        }
    };

    public Bitmap imageToBitmap(Image image) {
        // Get the buffer from the image
        ByteBuffer buffer = image.getPlanes()[0].getBuffer();
        byte[] bytes = new byte[buffer.remaining()];
        buffer.get(bytes);

        // Decode the byte array to a Bitmap
        return BitmapFactory.decodeByteArray(bytes, 0, bytes.length);
    }

    public Bitmap rotateBitmap(Bitmap bitmap, int rotationDegrees) {
        Matrix matrix = new Matrix();
        matrix.postRotate(rotationDegrees);
        return Bitmap.createBitmap(bitmap, 0, 0, bitmap.getWidth(), bitmap.getHeight(), matrix, false);
    }

    @Override
    protected void onResume() {
        super.onResume();
        startBackgroundThread();
    }

    @Override
    protected void onPause() {
        stopBackgroundThread();
        super.onPause();
    }

    private void startBackgroundThread() {
        backgroundThread = new HandlerThread("CameraBackground");
        backgroundThread.start();
        backgroundHandler = new Handler(backgroundThread.getLooper());
    }

    private void stopBackgroundThread() {
        backgroundThread.quitSafely();
        try {
            backgroundThread.join();
            backgroundThread = null;
            backgroundHandler = null;
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    private void copyFileFromAssetsToData(Context context, String fileName) throws IOException {
        AssetManager assetManager = context.getAssets();
        InputStream in = assetManager.open(fileName);
        File outFile = new File(context.getFilesDir(), fileName);
        FileOutputStream out = new FileOutputStream(outFile);

        byte[] buffer = new byte[1024];
        int read;
        while ((read = in.read(buffer)) != -1) {
            out.write(buffer, 0, read);
        }

        in.close();
        out.flush();
        out.close();
    }
}

File: face_emotion_recognition/ZeticMLangeFaceEmotionRecognition-Android/app/src/main/java/com/zeticai/faceemotionrecognition/MainActivity.kt
package com.zeticai.faceemotionrecognition

import android.Manifest
import android.content.pm.PackageManager
import android.os.Bundle
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts.RequestPermission
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.zetic.ZeticMLangeFeature.ZeticMLangeFeatureCameraController
import com.zetic.ZeticMLangeFeature.type.Box
import com.zeticai.faceemotionrecognition.feature.FaceDetection
import com.zeticai.faceemotionrecognition.feature.FaceEmotionRecognition
import kotlin.math.min
import kotlin.math.max

class MainActivity : AppCompatActivity() {
    private val requestPermissionLauncher =
        registerForActivityResult(RequestPermission()) { isGranted: Boolean ->
            if (isGranted)
                cameraController.startPreview()
            else
                Toast.makeText(this, "Camera Permission Not Granted!", Toast.LENGTH_SHORT).show()
        }

    private val visualizationSurfaceView: VisualizationSurfaceView by lazy { findViewById(R.id.visualizationSurfaceView) }
    private val zeticMLangeFeatureCameraController: ZeticMLangeFeatureCameraController = ZeticMLangeFeatureCameraController()
    private val faceDetection by lazy { FaceDetection(this, "face_detection_short_range") }
    private val faceEmotionRecognition by lazy { FaceEmotionRecognition(this, "face_emotion_recognition") }

    private val cameraController by lazy {
        CameraController(this,
            findViewById(R.id.surfaceView),
            findViewById(R.id.visualizationSurfaceView),
            { image, width, height ->
                processImage(image, width, height)
            },
            {
                zeticMLangeFeatureCameraController.setSurface(it)
            })
    }

    private fun clamp(a: Float, b: Float): Float {
        return max(0f, min(a, b))
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
    }

    override fun onResume() {
        super.onResume()

        if (ContextCompat.checkSelfPermission( this, Manifest.permission.CAMERA) == PackageManager.PERMISSION_GRANTED) {
            cameraController.startPreview()
        } else if (ActivityCompat.shouldShowRequestPermissionRationale(this, Manifest.permission.CAMERA)) {
            Toast.makeText(this, "Camera Permission Not Granted!", Toast.LENGTH_SHORT).show()
        } else {
            requestPermissionLauncher.launch(Manifest.permission.CAMERA)
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        cameraController.close()
        faceDetection.close()
        faceEmotionRecognition.close()
    }

    private fun processImage(image: ByteArray, width: Int, height: Int) {
        val imagePtr = zeticMLangeFeatureCameraController.frame(image)

        val faceDetectionResult = faceDetection.run(imagePtr)

        val faceEmotionRecognitionResult = if (faceDetectionResult.faceDetections.isEmpty()) null
        else {
            val res = faceDetectionResult.faceDetections[0]
            val resizeFactor = 0.2f
            val roi = Box(
                clamp(res.bbox.xMin * (1 - resizeFactor), 1f),
                clamp(res.bbox.yMin * (1 - resizeFactor), 1f),
                clamp(res.bbox.xMax * (1 + resizeFactor), 1f),
                clamp(res.bbox.yMax * (1 + resizeFactor), 1f),
            )
            faceEmotionRecognition.run(imagePtr, roi)
        }

        visualizationSurfaceView.visualize(
            faceDetectionResult,
            faceEmotionRecognitionResult)
    }

    companion object {
        const val TAG = "Main Activity"
    }
}
